nixos/modules/services/networking/tun2socks.nix

```
@@ -0,0 +1,98 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.tun2socks;
+  settingsFormat = pkgs.formats.yaml { };
+in
+{
+  options = {
+    services.tun2socks = {
+      enable = lib.mkEnableOption "tun2socks daemon, allow to use proxies through virtual TUN interface";
+      user = lib.mkOption {
+        type = lib.types.str;
+        default = "tun2socks";
+        description = "User account under which tun2socks runs.";
+      };
+
+      group = lib.mkOption {
+        type = lib.types.str;
+        default = "tun2socks";
+        description = "Group account under which tun2socks runs.";
+      };
+
+      settings = lib.mkOption {
+        description = ''
+          Configuration for tun2socks, see <https://github.com/xjasonlyu/tun2socks/wiki#usage>
+          for more information.
+        '';
+        default = { };
+        example = lib.literalExpression ''
+          {
+            device = "eth0";
+            fwmark = 1;
+            proxy = "http://192.168.1.1:1080";
+          }
+        '';
+        type = lib.types.submodule {
+          freeformType = settingsFormat.type;
+          options = {
+            device = lib.mkOption {
+              type = lib.types.str;
+              example = "tun0";
+              description = "The tun device to use.";
+            };
+            proxy = lib.mkOption {
+              type = lib.types.str;
+              example = "socks5://192.168.1.1:1080";
+              description = "The proxy to use [protocol://]host[:port]";
+            };
+          };
+        };
+      };
+
+      configFile = lib.mkOption {
+        default = settingsFormat.generate "/var/lib/tun2socks/config.yaml" cfg.settings;
```

```suggestion
        default = settingsFormat.generate "config.yaml" cfg.settings;
```

path here is unneeded, a quick search: https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+settingsFormat.generate&type=code